module/VuFind/src/VuFind/Controller/Feature/OnlinePaymentTrait.php

```diff
@@ -0,0 +1,329 @@
+     * @param array     $fines  List of fines
+     * @param ViewModel $view   View
+     *
+     * @return void
+     */
+    protected function handleOnlinePayment(array $patron, array $fines, ViewModel $view): void
+    {
+        $view->onlinePaymentEnabled = false;
+        $sourceIls = $patron['__source'] ?? 'default';
+
+        if (!($user = $this->getUser())) {
+            $this->handleError('Could not get user');
+            return;
+        }
+
+        // Check if online payment configuration exists and is valid for the ILS driver
+        $onlinePaymentManager = $this->serviceLocator->get(OnlinePaymentManager::class);
+        $paymentConfig = $onlinePaymentManager->getAndValidateOnlinePaymentConfig($patron);
+        if (!$paymentConfig) {
+            $this->handleDebugMsg("No online payment ILS configuration for $sourceIls");
+            return;
+        }
+
+        $selectFees = $paymentConfig['selectFines'] ?? false;
+        $pay = $this->formWasSubmitted('pay-confirm');
+        $selectedIds = ($selectFees && $pay)
+            ? $this->getRequest()->getPost()->get('selectedIDS', [])
+            : null;
+        $paymentDetails = $onlinePaymentManager->getAndCheckOnlinePaymentDetails(
+            $patron,
+            $fines,
+            $selectedIds
+        );
+        if ($selectedIds && !$paymentDetails['fines']) {
+            $this->handleError("Fines to pay missing from ILS driver for $sourceIls");
+            return;
+        }
+
+        $view->onlinePayment = true;
+        $view->paymentHandler = $onlinePaymentManager->getHandlerName($sourceIls);
+        $view->serviceFee = $paymentConfig['serviceFee'] ?? 0;
+        $view->minimumFee = $paymentConfig['minimumFee'] ?? 0;
+        $view->payableOnline = $paymentDetails['amount'];
+        $view->payableTotal = $paymentDetails['amount'] + $view->serviceFee;
+        $view->payableOnlineCnt = count($paymentDetails['fines']);
+        $view->nonPayableFines = count($fines) != count($paymentDetails['fines']);
+        $view->registerPayment = false;
+        $view->selectFees = $selectFees;
+
+        $paymentService = $this->getDbService(\VuFind\Db\Service\PaymentServiceInterface::class);
+        $lastPayment = null;
+        $receiptEnabled = $paymentConfig['receipt'] ?? false;
+        if ($receiptEnabled) {
+            $lastPayment = $paymentService->getLastPaidPaymentForPatron($patron['cat_username']);
+        }
+        if (
+            $lastPayment
+            && $this->params()->fromQuery('paymentReceipt') === 'true'
+        ) {
+            $receipt = $this->serviceLocator->get(\VuFind\OnlinePayment\Receipt::class);
+            $data = $receipt->createReceiptPDF($lastPayment, $paymentConfig);
+            header('Content-Type: application/pdf');
+            header(
+                'Content-disposition: inline; filename="' .
+                addcslashes($data['filename'], '"') . '"'
+            );
+            echo $data['pdf'];
```

Comment:
I generally try to avoid hard-coded `exit` calls in controllers, preferring to leverage the response object, redirect helper, etc. Maybe all of that is off the table with the impending exit of laminas-mvc, but I still wonder if we should try to be more formal, or at least encapsulate some of this logic outside of controller-land.